rl/agents/sarsa.py

```diff
@@ -0,0 +1,471 @@
+            state1_batch = [observation]
+
+            # Prepare and validate parameters.
+            state0_batch = self.process_state_batch(state0_batch)
+            state1_batch = self.process_state_batch(state1_batch)
+            terminal1_batch = np.array(terminal1_batch)
+            reward_batch = np.array(reward_batch)
+            assert reward_batch.shape == (1,)
+            assert terminal1_batch.shape == reward_batch.shape
+            assert len(action_batch) == len(reward_batch)
+
+            batch = self.process_state_batch(state1_batch)
+            # assert batch is not None
+            # print batch
+            q_values = self.compute_q_values(batch)
+            q_values = q_values.reshape((1, self.nb_actions))
+
+            action = [self.policy.select_action(q_values=q_values[0, :])]
+            if self.processor is not None:
+                action = self.processor.process_action(action)
+
+            self.nextAction = action[0]
+            q_batch = q_values[0, action]
+
+            assert q_batch.shape == (1,)
+            targets = np.zeros((1, self.nb_actions))
+            dummy_targets = np.zeros((1,))
+            masks = np.zeros((1, self.nb_actions))
+
+            # Compute r_t + gamma * Q(s_t+1, a_t+1)
+            discounted_reward_batch = self.gamma * q_batch
+            # Set discounted reward to zero for all states that were terminal.
+            discounted_reward_batch *= terminal1_batch
+            assert discounted_reward_batch.shape == reward_batch.shape
+            Rs = reward_batch + discounted_reward_batch
+            for idx, (target, mask, R, action) in enumerate(zip(targets, masks, Rs, action_batch)):
+                target[action] = R  # update action with estimated accumulated reward
+                dummy_targets[idx] = R
+                mask[action] = 1.  # enable loss for this specific action
+            targets = np.array(targets).astype('float32')
+            masks = np.array(masks).astype('float32')
+
+            # Finally, perform a single update on the entire batch. We use a dummy target since
+            # the actual loss is computed in a Lambda layer that needs more complex input. However,
+            # it is still useful to know the actual target to compute metrics properly.
+            state0_batch = state0_batch.reshape((1,) + state0_batch.shape)
+            ins = [state0_batch] if type(self.model.input) is not list else state0_batch
+            metrics = self.trainable_model.train_on_batch(ins + [targets, masks], [dummy_targets, targets])
+            metrics = [metric for idx, metric in enumerate(metrics) if idx not in (1, 2)]  # throw away individual losses
+            metrics += self.policy.metrics
+            if self.processor is not None:
+                metrics += self.processor.metrics
+
+        return metrics
+
+    def test(self, env, nb_episodes=1, action_repetition=1, callbacks=None, visualize=True,
```

Comment:
We really shouldn't duplicate the entire `test` method from `Agent`. What exactly has changed here? SARSA should easily fit into the general framework.